src/map/storage.c

```diff
@@ -96,37 +99,115 @@ static void do_reconnect_storage(void)
 	gstorage->db->foreach(gstorage->db, storage->reconnect_sub);
 }
 
+/**
+ * Get a storage id by its name (through @commands etc...)
+ * 
+ * @param[in] storage_name     pointer to the storage name char array.
+ * @return id of the storage or 0 if not found.
+ */
+int storage_get_id_by_name(const char* storage_name)
+{
+	nullpo_ret(storage_name);
+	int i = 0;
+
+	ARR_FIND(0, VECTOR_LENGTH(storage->configuration), i, strcmp(VECTOR_INDEX(storage->configuration, i).name, storage_name) == 0);
+
+	if (i == VECTOR_LENGTH(storage->configuration))
+		return 0;
+
+	return VECTOR_INDEX(storage->configuration, i).uid;
+}
+
+/**
+ * Get storage with a particular ID from a player.
+ * 
+ * @param[in] sd         pointer to map session data.
+ * @param[in] storage_id ID of the storage to receive.
+ * @return pointer to player's storage data structure or null if not found.
+ */
+struct storage_data* storage_ensure(struct map_session_data* sd, int storage_id)
+{
+	nullpo_retr(NULL, sd);
+	
+	if (storage_id <= 0)
+		return NULL;
+
+	int i = 0;
+	struct storage_data* stor = NULL;
+
+	ARR_FIND(0, VECTOR_LENGTH(sd->storage.list), i, (stor = &VECTOR_INDEX(sd->storage.list, i)) != NULL && stor->uid == storage_id);
+
+	if (i == VECTOR_LENGTH(sd->storage.list)) {
+		struct storage_data t_stor = { 0 };
+
+		t_stor.uid = storage_id;
+		VECTOR_INIT(t_stor.item);
+		VECTOR_ENSURE(sd->storage.list, 1, 1);
+		VECTOR_PUSH(sd->storage.list, t_stor);
+		stor = &VECTOR_LAST(sd->storage.list);
+	}
+
+	return stor;
+}
+
+/**
+ * Get a storage's settings through its ID.
+ * 
+ * @param[in] storage_id   the ID of the storage to find.
+ * @return storage settings of the storage in question.
+ */
+const struct storage_settings* storage_get_settings(int storage_id)
+{
+	if (storage_id <= 0)
+		return NULL;
+
+	int i = 0;
+
+	ARR_FIND(0, VECTOR_LENGTH(storage->configuration), i, VECTOR_INDEX(storage->configuration, i).uid == storage_id);
+
+	struct storage_settings* tmp_stor = NULL;
+	if (i < VECTOR_LENGTH(storage->configuration))
+		tmp_stor = &VECTOR_INDEX(storage->configuration, i);
+
+	return tmp_stor;
+}
+
 /*==========================================
  * Opens a storage. Returns:
  * 0 - success
  * 1 - fail
  *------------------------------------------*/
-static int storage_storageopen(struct map_session_data *sd)
+static int storage_storageopen(struct map_session_data *sd, struct storage_data *stor, enum storage_access_modes mode)
 {
-	nullpo_ret(sd);
+	nullpo_retr(1, sd);
+	nullpo_retr(1, stor);
 
-	if (sd->state.storage_flag != STORAGE_FLAG_CLOSED)
-		return 1; //Already open?
+	const struct storage_settings *stst = storage->get_settings(stor->uid);
+	nullpo_retr(1, stst);
 
-	if (sd->storage.received == false) {
-		clif->message(sd->fd, msg_sd(sd, MSGTBL_STORAGE_NOT_LOADED)); // Storage has not been loaded yet.
-		return 1;
-	}
```

Comment:
is removing the received check intended? from Haru message, he asked whether turning the message into assertion was intended (as an assertion doesn't inform the user why this happened).

But now, we would allow a not received storage to be "open".